File: tests/gpu_session_ffi/SessionProbe.swift
import Cera
import Foundation

func require(_ value: Bool, _ message: String = "Assertion failed") {
    precondition(value, message)
}

// Each call owns one complete conversation. ARC releases its Session on return.
func generateConversation(engine: CeraEngine, tokens: [UInt32]) throws -> [UInt32] {
    let session = try engine.newSession(config: SessionConfig(seed: 42))
    try session.appendTokens(tokens: tokens)
    return try session.generate(opts: options()).tokens
}

func options() -> GenerateOpts {
    GenerateOpts(maxTokens: 3, temperature: 0, ignoreEos: true, flushEveryTokens: 1)
}

func busy(_ engine: CeraEngine, _ config: SessionConfig) throws {
    do {
        _ = try engine.newSession(config: config)
        preconditionFailure("Expected Busy while the GPU session is retained")
    } catch FfiError.Busy { }
}

func gpuCases(bytes: Data, backend: BackendPreference, compression: KvCompression) async throws -> [String] {
    let load = EngineConfig(contextSize: 64, backend: backend)
    let config = SessionConfig(kvCompression: compression, seed: 42)
    let engine = try await CeraEngine.fromBytesAsync(bytes: bytes, config: load)
    let controlEngine = try await CeraEngine.fromBytesAsync(bytes: bytes, config: load)
    var active: Session? = try engine.newSession(config: config)
    weak let witness = active
    let control = try controlEngine.newSession(config: config)
    try active!.appendTokens(tokens: [0, 1, 0])
    try control.appendTokens(tokens: [0, 1, 0])
    try busy(engine, config)
    try busy(engine, SessionConfig(kvCompression: .turboQuant(seed: 99, keys: true, values: true)))

    let position = active!.position()
    let extracted = try active!.hiddenStatesForTokens(tokens: [1, 0])
    require(extracted.count == 2 * 32 * 4 && active!.position() == position)
    try busy(engine, config)
    active!.cancel()
    try busy(engine, config)
    active!.clearCancel()
    var invalid = options()
    invalid.grammar = "root ::= ("
    do {
        _ = try active!.generate(opts: invalid)
        preconditionFailure("Malformed grammar accepted")
    } catch FfiError.GrammarParse { }
    try busy(engine, config)
    let actual = try active!.generate(opts: options())
    let expected = try control.generate(opts: options())
    require(actual.tokens.count == 3 && actual.tokens == expected.tokens)
    require(active!.position() == control.position())

    try active!.reset()
    require(active!.position() == 0)
    try busy(engine, config)
    active = nil
    require(witness == nil, "Session wrapper did not release")
    do {
        _ = try engine.newSession(config: SessionConfig(
            kvCompression: .turboQuant(seed: 99, keys: true, values: true)))
        preconditionFailure("Compression conflict accepted")
    } catch FfiError.KvCompressionConflict { }
    let successor = try engine.newSession(config: config)
    try successor.appendTokens(tokens: [1, 0, 1])
    let freshEngine = try await CeraEngine.fromBytesAsync(bytes: bytes, config: load)
    let fresh = try freshEngine.newSession(config: config)
    try fresh.appendTokens(tokens: [1, 0, 1])
    require(try successor.generate(opts: options()).tokens == fresh.generate(opts: options()).tokens)

    var parent: CeraEngine? = try await CeraEngine.fromBytesAsync(bytes: bytes, config: load)
    weak let parentWitness = parent
    let retained = try parent!.newSession(config: config)
    parent = nil
    require(parentWitness == nil)
    try retained.appendTokens(tokens: [0, 1])
    require(try retained.generate(opts: options()).tokens.count == 3)
    return ["busy", "busy-before-config", "extraction-retains", "cancel-retains",
            "generation-error-retains", "continuation", "reset-retains", "release",
            "constructor-failure-release", "successor", "parent-release"]
}

final class PausedSink: ModalitySink, @unchecked Sendable {
    let entered = DispatchSemaphore(value: 0)
    let release = DispatchSemaphore(value: 0)
    private let lock = NSLock()
    private var paused = false
    private var done: [FinishReason] = []
    func onThoughtChunk(text: String) { }
    func onAudioFrames(pcm: [Float], sampleRate: UInt32) { }
    func onTextChunk(text: String) {
        lock.lock()
        let first = !paused
        paused = true
        lock.unlock()
        if first {
            entered.signal()
            require(release.wait(timeout: .now() + 20) == .success, "Callback release timed out")
        }
    }
    func onDone(reason: FinishReason) {
        lock.lock()
        done.append(reason)
        lock.unlock()
    }
    func waitForCallback() async -> Bool {
        await withCheckedContinuation { continuation in
            DispatchQueue.global().async {
                continuation.resume(returning: self.entered.wait(timeout: .now() + 20) == .success)
            }
        }
    }
    func completedReasons() -> [FinishReason] {
        lock.lock()
        defer { lock.unlock() }
        return done
    }
}

func launch(_ session: Session, _ sink: PausedSink) -> Task<GenerateSummary, Error> {
    Task.detached { try await session.generateStreamingAsync(opts: options(), sink: sink) }
}

func asyncCases(bytes: Data, backend: BackendPreference) async throws -> [String] {
    let engine = try await CeraEngine.fromBytesAsync(bytes: bytes, config: EngineConfig(contextSize: 64, backend: backend))
    var active: Session? = try engine.newSession(config: SessionConfig(seed: 42))
    weak let witness = active
    try active!.appendTokens(tokens: [0, 1])
    let sink = PausedSink()
    var work: Task<GenerateSummary, Error>? = launch(active!, sink)
    defer { sink.release.signal() }
    require(await sink.waitForCallback(), "No streaming callback")
    active!.cancel()
    active = nil
    require(witness != nil, "Pending call should retain the Swift Session")
    try busy(engine, SessionConfig())
    sink.release.signal()
    let summary = try await work!.value
    work = nil
    require(summary.finishReason == .cancelled)
    require(sink.completedReasons() == [.cancelled])
    require(witness == nil, "Completed call still retains the Session")
    let successor = try engine.newSession(config: SessionConfig(seed: 42))
    try successor.appendTokens(tokens: [1, 0])
    require(try successor.generate(opts: options()).tokens.count == 3)
    return ["async-retains", "async-cancel-release"]
}

@main
struct SessionProbe {
    static func main() async throws {
        let bytes = try Data(contentsOf: URL(fileURLWithPath: CommandLine.arguments[1]))
        var cases: [String] = []
        for (name, backend) in [("metal", BackendPreference.metal), ("wgpu", .gpu)] {
            for (mode, compression) in [("none", KvCompression.none),
                                        ("turboquant", .turboQuant(seed: 42, keys: true, values: true))] {
                cases += try await gpuCases(bytes: bytes, backend: backend, compression: compression).map { "\(name)/\(mode)/\($0)" }
            }
            cases += try await asyncCases(bytes: bytes, backend: backend).map { "\(name)/\($0)" }
            let engine = try await CeraEngine.fromBytesAsync(bytes: bytes, config: EngineConfig(contextSize: 64, backend: backend))
            let first = try generateConversation(engine: engine, tokens: [0, 1])
            let second = try generateConversation(engine: engine, tokens: [0, 1])
            require(first.count == 3 && first == second)
            cases.append("\(name)/scoped-example")
        }
        let cpu = try await CeraEngine.fromBytesAsync(bytes: bytes, config: EngineConfig(contextSize: 64, backend: .cpu))
        let first = try cpu.newSession(config: SessionConfig(seed: 42))
        let second = try cpu.newSession(config: SessionConfig(seed: 42))
        try first.appendTokens(tokens: [0, 1])
        try second.appendTokens(tokens: [0, 1])
        require(try first.generate(opts: options()).tokens == second.generate(opts: options()).tokens)
        cases.append("cpu/sharing")
        let json = try JSONSerialization.data(withJSONObject: ["cases": cases.sorted()])
        print(String(decoding: json, as: UTF8.self))
    }
}

File: tests/gpu_session_ffi/SessionProbe.kt
package sessionprobe

import java.nio.file.Files
import java.nio.file.Path
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking
import uniffi.cera_ffi.*

// Each call owns one complete conversation; use closes its Session on return.
fun generateConversation(engine: CeraEngine, tokens: List<UInt>): List<UInt> =
    engine.newSession(SessionConfig(seed = 42uL)).use { session ->
        session.appendTokens(tokens)
        session.generate(options()).tokens
    }

fun options() = GenerateOpts(maxTokens = 3u, temperature = 0f, ignoreEos = true, flushEveryTokens = 1u)

fun busy(engine: CeraEngine, config: SessionConfig) {
    try {
        engine.newSession(config).use { error("Expected Busy while the GPU session is retained") }
    } catch (_: FfiException.Busy) { }
}

suspend fun gpuCases(bytes: ByteArray, backend: BackendPreference, compression: KvCompression): List<String> {
    val opened = mutableListOf<AutoCloseable>()
    fun <T : AutoCloseable> own(value: T): T { opened.add(value); return value }
    try {
        val load = EngineConfig(contextSize = 64uL, backend = backend)
        val config = SessionConfig(kvCompression = compression, seed = 42uL)
        val engine = own(CeraEngine.fromBytesAsync(bytes, load))
        val controlEngine = own(CeraEngine.fromBytesAsync(bytes, load))
        val active = own(engine.newSession(config))
        val control = own(controlEngine.newSession(config))
        active.appendTokens(listOf(0u, 1u, 0u))
        control.appendTokens(listOf(0u, 1u, 0u))
        busy(engine, config)
        busy(engine, SessionConfig(kvCompression = KvCompression.TurboQuant(99uL, true, true)))
        val position = active.position()
        check(active.hiddenStatesForTokens(listOf(1u, 0u)).size == 2 * 32 * 4)
        check(active.position() == position)
        busy(engine, config)
        active.cancel()
        busy(engine, config)
        active.clearCancel()
        try {
            active.generate(options().copy(grammar = "root ::= ("))
            error("Malformed grammar accepted")
        } catch (_: FfiException.GrammarParse) { }
        busy(engine, config)
        val actual = active.generate(options())
        val expected = control.generate(options())
        check(actual.tokens.size == 3 && actual.tokens == expected.tokens)
        check(active.position() == control.position())
        active.reset()
        check(active.position() == 0u)
        busy(engine, config)
        active.close()
        try {
            engine.newSession(SessionConfig(kvCompression = KvCompression.TurboQuant(99uL, true, true))).use {
                error("Compression conflict accepted")
            }
        } catch (_: FfiException.KvCompressionConflict) { }
        val successor = own(engine.newSession(config))
        successor.appendTokens(listOf(1u, 0u, 1u))
        val freshEngine = own(CeraEngine.fromBytesAsync(bytes, load))
        val fresh = own(freshEngine.newSession(config))
        fresh.appendTokens(listOf(1u, 0u, 1u))
        check(successor.generate(options()).tokens == fresh.generate(options()).tokens)
        val parent = own(CeraEngine.fromBytesAsync(bytes, load))
        val retained = own(parent.newSession(config))
        parent.close()
        retained.appendTokens(listOf(0u, 1u))
        check(retained.generate(options()).tokens.size == 3)
        return listOf("busy", "busy-before-config", "extraction-retains", "cancel-retains",
            "generation-error-retains", "continuation", "reset-retains", "release",
            "constructor-failure-release", "successor", "parent-release")
    } finally {
        opened.asReversed().forEach { it.close() }
    }
}

class PausedSink : ModalitySink {
    val entered = CountDownLatch(1)
    val release = CountDownLatch(1)
    private val paused = AtomicBoolean(false)
    private val done = mutableListOf<FinishReason>()
    override fun onThoughtChunk(text: String) { }
    override fun onAudioFrames(pcm: List<Float>, sampleRate: UInt) { }
    override fun onTextChunk(text: String) {
        if (paused.compareAndSet(false, true)) {
            entered.countDown()
            check(release.await(20, TimeUnit.SECONDS)) { "Callback release timed out" }
        }
    }
    override fun onDone(reason: FinishReason) { synchronized(done) { done.add(reason) } }
    fun completedReasons(): List<FinishReason> = synchronized(done) { done.toList() }
}

suspend fun asyncCases(bytes: ByteArray, backend: BackendPreference): List<String> = coroutineScope {
    CeraEngine.fromBytesAsync(bytes, EngineConfig(contextSize = 64uL, backend = backend)).use { engine ->
        val active = engine.newSession(SessionConfig(seed = 42uL))
        val sink = PausedSink()
        try {
            active.appendTokens(listOf(0u, 1u))
            val work = async(Dispatchers.Default) { active.generateStreamingAsync(options(), sink) }
            try {
                check(sink.entered.await(20, TimeUnit.SECONDS)) { "No streaming callback" }
                active.cancel()
                active.close()
                busy(engine, SessionConfig())
            } finally {
                sink.release.countDown()
            }
            val summary = work.await()
            check(summary.finishReason == FinishReason.Cancelled)
            check(sink.completedReasons() == listOf(FinishReason.Cancelled))
            engine.newSession(SessionConfig(seed = 42uL)).use { successor ->
                successor.appendTokens(listOf(1u, 0u))
                check(successor.generate(options()).tokens.size == 3)
            }
        } finally {
            sink.release.countDown()
            active.close()
        }
    }
    listOf("async-retains", "async-cancel-release")
}

fun main(args: Array<String>) = runBlocking {
    val bytes = Files.readAllBytes(Path.of(args.single()))
    val cases = mutableListOf<String>()
    for ((name, backend) in listOf("metal" to BackendPreference.METAL, "wgpu" to BackendPreference.GPU)) {
        for ((mode, compression) in listOf("none" to KvCompression.None,
            "turboquant" to KvCompression.TurboQuant(42uL, true, true))) {
            cases += gpuCases(bytes, backend, compression).map { "$name/$mode/$it" }
        }
        cases += asyncCases(bytes, backend).map { "$name/$it" }
        CeraEngine.fromBytesAsync(bytes, EngineConfig(contextSize = 64uL, backend = backend)).use { engine ->
            val first = generateConversation(engine, listOf(0u, 1u))
            check(first.size == 3 && first == generateConversation(engine, listOf(0u, 1u)))
        }
        cases += "$name/scoped-example"
    }
    CeraEngine.fromBytesAsync(bytes, EngineConfig(contextSize = 64uL, backend = BackendPreference.CPU)).use { cpu ->
        cpu.newSession(SessionConfig(seed = 42uL)).use { first ->
            cpu.newSession(SessionConfig(seed = 42uL)).use { second ->
                first.appendTokens(listOf(0u, 1u))
                second.appendTokens(listOf(0u, 1u))
                check(first.generate(options()).tokens == second.generate(options()).tokens)
            }
        }
    }
    cases += "cpu/sharing"
    println(cases.sorted().joinToString(prefix = "{\"cases\":[", postfix = "]}") { "\"$it\"" })
}
